gtsam/nonlinear/GncOptimizer.h

```diff
@@ -0,0 +1,459 @@
+        && lossType == other.lossType && maxIterations == other.maxIterations
+        && std::fabs(barcSq - other.barcSq) <= tol
+        && std::fabs(muStep - other.muStep) <= tol
+        && verbosityGNC == other.verbosityGNC
+        && knownInliers == other.knownInliers;
+  }
+  /// print function
+  void print(const std::string& str) const {
+    std::cout << str << "\n";
+    switch (lossType) {
+    case GM:
+      std::cout << "lossType: Geman McClure" << "\n";
+      break;
+    case TLS:
+      std::cout << "lossType: Truncated Least-squares" << "\n";
+      break;
+    default:
+      throw std::runtime_error("GncParams::print: unknown loss type.");
+    }
+    std::cout << "maxIterations: " << maxIterations << "\n";
+    std::cout << "barcSq: " << barcSq << "\n";
+    std::cout << "muStep: " << muStep << "\n";
+    std::cout << "relativeCostTol: " << relativeCostTol << "\n";
+    std::cout << "weightsTol: " << weightsTol << "\n";
+    std::cout << "verbosityGNC: " << verbosityGNC << "\n";
+    for (size_t i = 0; i < knownInliers.size(); i++)
+      std::cout << "knownInliers: " << knownInliers[i] << "\n";
+    baseOptimizerParams.print(str);
+  }
+};
+
+/* ************************************************************************* */
+template<class GncParameters>
+class GncOptimizer {
+public:
+  /** For each parameter, specify the corresponding optimizer: e.g., GaussNewtonParams -> GaussNewtonOptimizer */
+  typedef typename GncParameters::OptimizerType BaseOptimizer;
+
+private:
+  NonlinearFactorGraph nfg_;
+  Values state_;
+  GncParameters params_;
+  Vector weights_; // this could be a local variable in optimize, but it is useful to make it accessible from outside
+
+public:
+  /// Constructor
+  GncOptimizer(const NonlinearFactorGraph& graph, const Values& initialValues,
+      const GncParameters& params = GncParameters()) :
+      state_(initialValues), params_(params) {
+
+    // make sure all noiseModels are Gaussian or convert to Gaussian
+    nfg_.resize(graph.size());
+    for (size_t i = 0; i < graph.size(); i++) {
+      if (graph[i]) {
+        NoiseModelFactor::shared_ptr factor = boost::dynamic_pointer_cast<
+            NoiseModelFactor>(graph[i]);
+        noiseModel::Robust::shared_ptr robust = boost::dynamic_pointer_cast<
+            noiseModel::Robust>(factor->noiseModel());
+        if (robust) { // if the factor has a robust loss, we have to change it:
+          SharedNoiseModel gaussianNoise = robust->noise();
+          NoiseModelFactor::shared_ptr gaussianFactor =
+              factor->cloneWithNewNoiseModel(gaussianNoise);
+          nfg_[i] = gaussianFactor;
+        } else { // else we directly push it back
+          nfg_[i] = factor;
+        }
+      }
+    }
+  }
+
+  /// Access a copy of the internal factor graph
+  NonlinearFactorGraph getFactors() const {
+    return NonlinearFactorGraph(nfg_);
+  }
+  /// Access a copy of the internal values
```

Comment:
Here and below: add return. Make sure to format with clang-format or cpp-format w Google style.